### PR TITLE
Specifies the new KerasCV bounding box format

### PR DESCRIPTION
**Pull request by LukeWood**
This PR Specifies a new KerasCV bounding box format based on dictionaries.  Unlike the previous version, we now rely on an explicit value for the `mask` instead of simply relying on the class ID.

The new format is defined as follows:

```
For a set of bounding boxes to be valid it must satisfy the following conditions:
- `bounding_boxes` must be a dictionary
- contains keys `"boxes"` and `"classes"`
- each entry must have matching first two dimensions; representing the batch axis
    and the number of boxes per image axis.
- either both `"boxes"` and `"classes"` are batched, or both are unbatched

Additionally, one of the following must be satisfied:
- "boxes" and "classes" are both Ragged
- `"boxes"` and `"classes"` are dense and `"mask"` is a Tensor, matching the
    dimensionality of `"classes"`
- `"boxes"` and `"classes"` are dense and `"mask"` is set to `False`
- `"boxes"` and `"classes"` are unbatched
```

---

Still working on unit tests, mainly looking for agreement on the format specification

### CODE REVIEW

**Review by ianstenbit — CHANGES_REQUESTED** [review]

File: keras_cv/bounding_box/validate.py
Comment (LukeWood) [reply keras_cv/bounding_box/validate.py:18]: So, personally I think this is better:

- it is more explicit
- it gives us a way to distinguish padded bounding box tensors (sometimes even non-Ragged tensors are not padded)
- it allows us flexibility in the format of "classes" going forwards.  For example, what if we want a OHE class label?  Once we OHE we can actually label smooth, apply true mixup, etc.

Comment (LukeWood) [reply keras_cv/bounding_box/validate.py:18]: Open to feedback and thoughts on this!

**Comment by martin-gorner** [comment]
A couple of design questions:
* have you considered using a flat dictionary instead of a nested dictionary ?
```
{"images":..., "bounding_boxes": ..., "classes": ..., "bb_mask": ...}
#instead of:
{"images":..., "bounding_boxes":{"boxes": ..., "classes": ..., "mask": ...}}
```
Not sure which one is better.
* Now that we have a dictionary, should we include a "bounding_box_format" field ?
* see other Q about masking on internal chat

**Comment by LukeWood** [comment]
> A couple of design questions:
> 
> * have you considered using a flat dictionary instead of a nested dictionary ?
> 
> ```
> {"images":..., "bounding_boxes": ..., "classes": ..., "bb_mask": ...}
> #instead of:
> {"images":..., "bounding_boxes":{"boxes": ..., "classes": ..., "mask": ...}}
> ```
> 
> Not sure which one is better.
> 
> * Now that we have a dictionary, should we include a "bounding_box_format" field ?
> * see other Q about masking on internal chat

Yeah I did consider the flat dictionary, but I do think with multiple label types (ie. segmentation maps) this will get too complicated too fast.

**Comment by LukeWood** [comment]
> A couple of design questions:
> 
> * have you considered using a flat dictionary instead of a nested dictionary ?
> 
> ```
> {"images":..., "bounding_boxes": ..., "classes": ..., "bb_mask": ...}
> #instead of:
> {"images":..., "bounding_boxes":{"boxes": ..., "classes": ..., "mask": ...}}
> ```
> 
> Not sure which one is better.
> 
> * Now that we have a dictionary, should we include a "bounding_box_format" field ?
> * see other Q about masking on internal chat

re: Masked Tensors, lets wait for GA before we depend on something like that.

File: keras_cv/bounding_box/validate.py
Comment (ianstenbit) [inline keras_cv/bounding_box/validate.py:5]: nit: complies with

**Review by ianstenbit — APPROVED** [review]
I am fine with this approach.

I would prefer that `mask` is optional, and that -1-padded dense tensors are automagically ignored, but perhaps that gets too ugly in terms of implementation.

Let's wait until we have LGTM from all before merging

**Pull request closed without merging**
